crypto/evp/p_lib.c

```diff
@@ -1335,23 +1335,21 @@ int EVP_PKEY_get_group_name(const EVP_PKEY *pkey, char *gname, size_t gname_sz,
                                           gname, gname_sz, gname_len);
 }
 
-int EVP_PKEY_supports_digest_nid(EVP_PKEY *pkey, int nid)
+int EVP_PKEY_digestsign_supports_digest(EVP_PKEY *pkey, OSSL_LIB_CTX *libctx,
+                                        const char *name, const char *propq)
 {
-    int rv, default_nid;
+    int rv;
+    EVP_MD_CTX *ctx = NULL;
 
-    rv = evp_pkey_asn1_ctrl(pkey, ASN1_PKEY_CTRL_SUPPORTS_MD_NID, nid, NULL);
-    if (rv == -2) {
-        /*
-         * If there is a mandatory default digest and this isn't it, then
-         * the answer is 'no'.
-         */
-        rv = EVP_PKEY_get_default_digest_nid(pkey, &default_nid);
-        if (rv == 2)
-            return (nid == default_nid);
-        /* zero is an error from EVP_PKEY_get_default_digest_nid() */
-        if (rv == 0)
-            return -1;
-    }
+    if ((ctx = EVP_MD_CTX_new()) == NULL)
+        return -1;
+
+    ERR_set_mark();
+    rv = EVP_DigestSignInit_ex(ctx, NULL, name, libctx,
```

Comment:
Is this equivalent to what it did before?
ASN1_PKEY_CTRL_DEFAULT_MD_NID for ec for example returned SM3 or SHA256..
rsa pss returned md?

Reply:
We will have to decide to support, or not, signatures that combine those keytypes with those digests.
I *assume* that if `ASN1_PKEY_CTRL_DEFAULT_MD_NID` returned a certain NID, `EVP_DigestSignIni` using a digest produced with that NID wouldn't refuse it.

So, if it's not equivalent, we either need to restore support for the "offending" digest, or document the breaking change.